ssl/statem/statem_clnt.c

```diff
@@ -3336,7 +3336,12 @@ int tls_construct_client_certificate(SSL *s, WPACKET *pkt)
                     SSL3_CC_HANDSHAKE | SSL3_CHANGE_CIPHER_CLIENT_WRITE))) {
         SSLerr(SSL_F_TLS_CONSTRUCT_CLIENT_CERTIFICATE,
                SSL_R_CANNOT_CHANGE_CIPHER);
-        goto err;
+        /*
+         * This is a fatal error, which leaves
+         * enc_write_ctx in an inconsistent state
+         * and thus ssl3_send_alert may crash.
+         */
+        return 0;
```

Comment:
I mean #2638:

In an ideal world the SSL object would return SSL_ERROR_WANT_WRITE
until the alert is sent and then SSL_ERROR_SSL, but if the error was OOM while
allocating an internal buffer, it would return SSL_ERROR_SSL immediately,
and ssl3_send_alert would known internally when it is safe to use the enc_write_ctx.

In a less perfect world, we return -1 immediately if we get out of memory,
and do not try to send an alert about it.